### PR TITLE
Handle service exceptions

### PR DESCRIPTION
**Pull request by sadevelopment**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@teamhubapp/firebase-multiplatform",
-  "version": "0.1.0",
+  "version": "0.1.5",
```
Comment (nbransby) [inline package.json]: needs minor bump not patch as it adds new apis

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt [hunk 2]
```diff
@@ -112,3 +123,53 @@ internal fun fromJson(data: Any?, valueType: KClass<*>? = null): Any? = when(dat
...
 
+
+internal suspend fun <T, R> T.runActualWithHandler(function: suspend T.() -> R): R {
+    var exception: Exception?
+
+    try {
+        return function()
+    } catch(e: dynamic) {
+        if(e.code) {
+            exception = when(e.code) {
```
Comment (nbransby) [inline src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: just throw result of the when

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt [hunk 3]
```diff
@@ -112,3 +123,53 @@ internal fun fromJson(data: Any?, valueType: KClass<*>? = null): Any? = when(dat
...
+                "resource-exhausted" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.RESOURCE_EXHAUSTED)
+                "failed-precondition" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.FAILED_PRECONDITION)
+                "aborted" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.ABORTED)
+                "out-of-range" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.OUT_OF_RANGE)
+                "unimplemented" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.UNIMPLEMENTED)
+                "internal" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.INTERNAL)
+                "unavailable" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.UNAVAILABLE)
+                "data-loss" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.DATA_LOSS)
+                "unauthenticated" -> FirebaseFirestoreException(e.message as String, FirestoreExceptionCode.UNAUTHENTICATED)
```
Comment (nbransby) [inline src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: is there more that just code you can test to see if this error is really a firestore error? maybe theres another attribute to identity it as such?

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt [hunk 4]
```diff
@@ -112,3 +123,53 @@ internal fun fromJson(data: Any?, valueType: KClass<*>? = null): Any? = when(dat
...
+
+                else -> FirebaseException(e)
+            }
+        } else if(e is DatabaseException) {
```
Comment (nbransby) [inline src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: how can this work?

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt [hunk 5]
```diff
@@ -112,3 +123,53 @@ internal fun fromJson(data: Any?, valueType: KClass<*>? = null): Any? = when(dat
...
+            exception = e
+        } else {
+            exception = FirebaseException(e)
```
Comment (nbransby) [inline src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: why?

File: package.json (hunk 1 above)
Comment (sadevelopment) [reply package.json]: Bumped to 0.2.0

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt (hunk 2 above)
Comment (sadevelopment) [reply src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: Changed to throw when

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt (hunk 3 above)
Comment (sadevelopment) [reply src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: The is only 1 other attribute and that is name = "FirebaseError"

Comment (nbransby) [reply src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: gonna put a test in for that too then?

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt (hunk 4 above)
Comment (sadevelopment) [reply src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: I removed this .... forgot its an Error

File: src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt (hunk 5 above)
Comment (sadevelopment) [reply src/jsMain/kotlin/multiplatform/com/google/firebase/firebase.kt]: General catch all for exceptions that don't contain a code attribute

**Pull request merged**